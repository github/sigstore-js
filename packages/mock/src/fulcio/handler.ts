/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import assert from 'assert';
import { generateKeyPairSync } from 'crypto';
import * as jose from 'jose';
import type { Handler, HandlerFn, HandlerFnResult } from '../shared.types';
import type { CA, ExtensionValue } from './ca';

const CREATE_SIGNING_CERT_PATH = '/api/v2/signingCert';
const DEFAULT_SUBJECT = 'NO-SUBJECT';
const DEFAULT_ISSUER = 'https://fake.oidcissuer.com';

const ISSUER_EXT_OID_V1 = '1.3.6.1.4.1.57264.1.1';
const GH_WORKFLOW_TRIGGER_EXT_OID = '1.3.6.1.4.1.57264.1.2';
const GH_WORKFLOW_SHA_EXT_OID = '1.3.6.1.4.1.57264.1.3';
const GH_WORKFLOW_NAME_EXT_OID = '1.3.6.1.4.1.57264.1.4';
const GH_WORKFLOW_REPO_EXT_OID = '1.3.6.1.4.1.57264.1.5';
const GH_WORKFLOW_REF_EXT_OID = '1.3.6.1.4.1.57264.1.6';
const ISSUER_EXT_OID_V2 = '1.3.6.1.4.1.57264.1.8';
const BUILD_SIGNER_URI_EXT_OID = '1.3.6.1.4.1.57264.1.9';
const BUILD_SIGNER_DIGEST_EXT_OID = '1.3.6.1.4.1.57264.1.10';
const RUNNER_ENVIRONMENT_EXT_OID = '1.3.6.1.4.1.57264.1.11';
const SOURCE_REPO_URI_EXT_OID = '1.3.6.1.4.1.57264.1.12';
const SOURCE_REPO_DIGEST_EXT_OID = '1.3.6.1.4.1.57264.1.13';
const SOURCE_REPO_REF_EXT_OID = '1.3.6.1.4.1.57264.1.14';
const SOURCE_REPO_ID_EXT_OID = '1.3.6.1.4.1.57264.1.15';
const SOURCE_REPO_OWNER_URI_EXT_OID = '1.3.6.1.4.1.57264.1.16';
const SOURCE_REPO_OWNER_ID_EXT_OID = '1.3.6.1.4.1.57264.1.17';
const BUILD_CONFIG_URI_EXT_OID = '1.3.6.1.4.1.57264.1.18';
const BUILD_CONFIG_DIGEST_EXT_OID = '1.3.6.1.4.1.57264.1.19';
const BUILD_TRIGGER_EXT_OID = '1.3.6.1.4.1.57264.1.20';
const RUN_INVOCATION_URI_EXT_OID = '1.3.6.1.4.1.57264.1.21';
const SOURCE_REPO_VISIBILITY_EXT_OID = '1.3.6.1.4.1.57264.1.22';

interface FulcioHandlerOptions {
  strict?: boolean;
  subjectClaim?: string;
}

export function fulcioHandler(
  ca: CA,
  opts: FulcioHandlerOptions = {}
): Handler {
  return {
    path: CREATE_SIGNING_CERT_PATH,
    fn: createSigningCertHandler(ca, opts),
  };
}

function createSigningCertHandler(
  ca: CA,
  opts: FulcioHandlerOptions
): HandlerFn {
  const strict = opts.strict ?? true;
  const subjectClaim = opts.subjectClaim || 'sub';

  return async (body: string): Promise<HandlerFnResult> => {
    try {
      // Extract relevant fields from the request
      const { subject, publicKey, claims } = strict
        ? parseBody(body, subjectClaim)
        : stubBody();

      const extensions = extensionFromClaims(claims);

      // Request certificate from CA
      const cert = await ca.issueCertificate({
        publicKey: fromPEM(publicKey),
        subjectAltName: subject,
        extensions: extensions,
      });

      // Format the response
      const response = buildResponse(cert, ca.rootCertificate);

      return { statusCode: 201, response, contentType: 'application/json' };
    } catch (e) {
      assert(e instanceof Error);
      return { statusCode: 400, response: e.message };
    }
  };
}

function parseBody(
  body: string,
  subjectClaim: string
): { subject: string; publicKey: string; claims: Record<string, string> } {
  const json = JSON.parse(body.toString());
  const oidc = json.credentials.oidcIdentityToken;
  const pem = json.publicKeyRequest.publicKey.content;

  // Decode the JWT
  const claims = jose.decodeJwt(oidc) as Record<string, string>;

  /* istanbul ignore next */
  return {
    subject: claims[subjectClaim] || DEFAULT_SUBJECT,
    publicKey: pem,
    claims: { iss: DEFAULT_ISSUER, ...claims },
  };
}

function stubBody(): {
  subject: string;
  publicKey: string;
  claims: Record<string, string>;
} {
  const { publicKey } = generateKeyPairSync('ec', {
    namedCurve: 'P-256',
  });
  return {
    subject: DEFAULT_SUBJECT,
    publicKey: publicKey.export({ format: 'pem', type: 'spki' }).toString(),
    claims: { iss: DEFAULT_ISSUER },
  };
}

function buildResponse(leaf: Buffer, root: Buffer): string {
  const body = {
    signedCertificateEmbeddedSct: {
      chain: {
        certificates: [toPEM(leaf), toPEM(root)],
      },
    },
  };
  return JSON.stringify(body);
}

function extensionFromClaims(claims: Record<string, string>): ExtensionValue[] {
  const extensions: ExtensionValue[] = [];
  const baseURL = 'https://github.com';

  for (const [key, value] of Object.entries(claims)) {
    switch (key) {
      case 'iss':
        extensions.push({
          oid: ISSUER_EXT_OID_V1,
          value: value,
          legacy: true,
        });
        extensions.push({ oid: ISSUER_EXT_OID_V2, value: value });
        break;
      case 'event_name':
        extensions.push({
          oid: GH_WORKFLOW_TRIGGER_EXT_OID,
          value: value,
          legacy: true,
        });
        extensions.push({ oid: BUILD_TRIGGER_EXT_OID, value: value });
        break;
      case 'sha':
        extensions.push({
          oid: GH_WORKFLOW_SHA_EXT_OID,
          value: value,
          legacy: true,
        });
        extensions.push({ oid: SOURCE_REPO_DIGEST_EXT_OID, value: value });
        break;
      case 'workflow':
        extensions.push({
          oid: GH_WORKFLOW_NAME_EXT_OID,
          value: value,
          legacy: true,
        });
        break;
      case 'repository':
        extensions.push({
          oid: GH_WORKFLOW_REPO_EXT_OID,
          value: value,
          legacy: true,
        });
        extensions.push({
          oid: SOURCE_REPO_URI_EXT_OID,
          value: `${baseURL}/${value}`,
        });
        break;
      case 'ref':
        extensions.push({
          oid: GH_WORKFLOW_REF_EXT_OID,
          value: value,
          legacy: true,
        });
        extensions.push({
          oid: SOURCE_REPO_REF_EXT_OID,
          value: value,
        });
        break;
      case 'job_workflow_ref':
        extensions.push({
          oid: BUILD_SIGNER_URI_EXT_OID,
          value: `${baseURL}/${value}`,
        });
        break;
      case 'job_workflow_sha':
        extensions.push({
          oid: BUILD_SIGNER_DIGEST_EXT_OID,
          value: value,
        });
        break;
      case 'runner_environment':
        extensions.push({
          oid: RUNNER_ENVIRONMENT_EXT_OID,
          value: value,
        });
        break;
      case 'repository_id':
        extensions.push({
          oid: SOURCE_REPO_ID_EXT_OID,
          value: value,
        });
        break;
      case 'repository_owner':
        extensions.push({
          oid: SOURCE_REPO_OWNER_URI_EXT_OID,
          value: `${baseURL}/${value}`,
        });
        break;
      case 'repository_owner_id':
        extensions.push({
          oid: SOURCE_REPO_OWNER_ID_EXT_OID,
          value: value,
        });
        break;
      case 'workflow_ref':
        extensions.push({
          oid: BUILD_CONFIG_URI_EXT_OID,
          value: `${baseURL}/${value}`,
        });
        break;
      case 'workflow_sha':
        extensions.push({
          oid: BUILD_CONFIG_DIGEST_EXT_OID,
          value: value,
        });
        break;
      case 'repository_visibility':
        extensions.push({
          oid: SOURCE_REPO_VISIBILITY_EXT_OID,
          value: value,
        });
        break;
    }
  }

  if (claims['repository'] && claims['run_id'] && claims['run_attempt']) {
    extensions.push({
      oid: RUN_INVOCATION_URI_EXT_OID,
      value: `${baseURL}/${claims['repository']}/actions/runs/${claims['run_id']}/attempts/${claims['run_attempt']}`,
    });
  }

  return extensions;
}

// PEM string to DER-encoded byte buffer conversion
function fromPEM(pem: string): Buffer {
  return Buffer.from(
    pem.replace(/-{5}(BEGIN|END) .*-{5}/gm, '').replace(/\s/gm, ''),
    'base64'
  );
}

// DER-encoded byte buffer to PEM string conversion
function toPEM(der: Buffer): string {
  return [
    '-----BEGIN CERTIFICATE-----',
    der.toString('base64'),
    '-----END CERTIFICATE-----',
    '',
  ].join('\n');
}
