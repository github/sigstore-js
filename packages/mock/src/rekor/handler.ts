/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import assert from 'assert';
import type { Handler, HandlerFn, HandlerFnResult } from '../shared.types';
import type { TLog } from './tlog';

const CREATE_ENTRY_PATH = '/api/v1/log/entries';

interface RekorHandlerOptions {
  strict?: boolean;
}

export function rekorHandler(
  tlog: TLog,
  opts: RekorHandlerOptions = {}
): Handler {
  return {
    path: CREATE_ENTRY_PATH,
    fn: createEntryHandler(tlog, opts),
  };
}

function createEntryHandler(tlog: TLog, opts: RekorHandlerOptions): HandlerFn {
  const strict = opts.strict ?? true;

  return async (body: string): Promise<HandlerFnResult> => {
    try {
      const proposedEntry = strict
        ? JSON.parse(body)
        : { kind: 'intoto', apiVersion: '0.0.2' };
      const tlogEntry = await tlog.log(proposedEntry);
      const response = JSON.stringify(tlogEntry);

      return { statusCode: 201, response, contentType: 'application/json' };
    } catch (e) {
      assert(e instanceof Error);
      return { statusCode: 400, response: e.message };
    }
  };
}
