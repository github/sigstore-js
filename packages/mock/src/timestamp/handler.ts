/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import assert from 'assert';
import type { Handler, HandlerFn, HandlerFnResult } from '../shared.types';
import type { TSA, TimestampRequest } from './tsa';

const OID_SHA256_ALGO_ID = '2.16.840.1.101.3.4.2.1';
const OID_SHA384_ALGO_ID = '2.16.840.1.101.3.4.2.2';
const OID_SHA512_ALGO_ID = '2.16.840.1.101.3.4.2.3';

const OID_TSA_POLICY = '1.3.6.1.4.1.57264.2';

const DEFAULT_NONCE = 1234567890;

const CREATE_TIMESTAMP_PATH = '/api/v1/timestamp';

interface TSAHandlerOptions {
  strict?: boolean;
}

export function tsaHandler(tsa: TSA, opts: TSAHandlerOptions = {}): Handler {
  return {
    path: CREATE_TIMESTAMP_PATH,
    fn: createTimestampHandler(tsa, opts),
  };
}

function createTimestampHandler(tsa: TSA, opts: TSAHandlerOptions): HandlerFn {
  const strict = opts.strict ?? true;

  return async (body: string): Promise<HandlerFnResult> => {
    try {
      const tsr = strict ? parseRequest(body) : stubRequest();
      const timestamp = await tsa.timestamp(tsr);

      return {
        statusCode: 201,
        response: timestamp,
        contentType: 'application/timestamp-reply',
      };
    } catch (e) {
      assert(e instanceof Error);
      return { statusCode: 400, response: e.message };
    }
  };
}

function parseRequest(body: string): TimestampRequest {
  const json = JSON.parse(body.toString());

  return {
    artifactHash: Buffer.from(json.artifactHash, 'base64'),
    hashAlgorithmOID: hashToOID(json.hashAlgorithm),
    nonce: json.nonce || DEFAULT_NONCE,
    policyOID: json.tsaPolicyOID || OID_TSA_POLICY,
    certReq: json.certificates ?? false,
  };
}

function stubRequest(): TimestampRequest {
  return {
    artifactHash: Buffer.from('deadbeef', 'hex'),
    hashAlgorithmOID: OID_SHA256_ALGO_ID,
    nonce: DEFAULT_NONCE,
    policyOID: OID_TSA_POLICY,
    certReq: false,
  };
}

function hashToOID(hash: string): string {
  switch (true) {
    /* istanbul ignore next */
    case /512/.test(hash):
      return OID_SHA512_ALGO_ID;
    /* istanbul ignore next */
    case /384/.test(hash):
      return OID_SHA384_ALGO_ID;
    default:
      return OID_SHA256_ALGO_ID;
  }
}
