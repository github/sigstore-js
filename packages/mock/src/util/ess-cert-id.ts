import * as asn1js from 'asn1js';
import * as pkijs from 'pkijs';
import * as pvutils from 'pvutils';

const CERT_HASH = 'certHash';
const ISSUER_SERIAL = 'issuerSerial';

interface IESSCertIDv2 {
  certHash: asn1js.OctetString;
  issuerSerial: pkijs.IssuerSerial;
}

type ESSCertIDv2Parameters = pkijs.PkiObjectParameters & Partial<IESSCertIDv2>;

// https://datatracker.ietf.org/doc/html/rfc5035#section-4
export class ESSCertIDv2 extends pkijs.PkiObject implements IESSCertIDv2 {
  public static override CLASS_NAME = 'ESSCertIDv2';

  public certHash!: asn1js.OctetString;
  public issuerSerial!: pkijs.IssuerSerial;

  constructor(parameters: ESSCertIDv2Parameters = {}) {
    super();

    this.certHash = pvutils.getParametersValue(
      parameters,
      CERT_HASH,
      new asn1js.OctetString()
    );
    this.issuerSerial = pvutils.getParametersValue(
      parameters,
      ISSUER_SERIAL,
      new pkijs.IssuerSerial()
    );
  }

  public override toSchema(): asn1js.Sequence {
    const result = new asn1js.Sequence({
      value: [this.certHash, this.issuerSerial.toSchema()],
    });

    return result;
  }

  /* istanbul ignore next */
  public override fromSchema(): void {
    throw new Error('Not implemented');
  }

  /* istanbul ignore next */
  public override toJSON(): IESSCertIDv2 {
    throw new Error('Not implemented');
  }
}
