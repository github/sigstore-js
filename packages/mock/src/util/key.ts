import { Crypto, CryptoKey } from '@peculiar/webcrypto';
import crypto from 'crypto';

export function generateKeyPair(
  namedCurve: string = 'P-256'
): crypto.KeyPairKeyObjectResult {
  return crypto.generateKeyPairSync('ec', { namedCurve });
}

// Translates a Node.js KeyObject to a WebCrypto CryptoKey
export async function keyObjectToCryptoKey(
  keyObject: crypto.KeyObject
): Promise<CryptoKey> {
  const keyData = keyObject.export({ format: 'jwk' });
  const algorithm = { name: 'ECDSA', namedCurve: namedCurve(keyObject) };
  const extractable = true;
  const usages: KeyUsage[] =
    keyObject.type === 'private' ? ['sign'] : ['verify'];

  return await new Crypto().subtle.importKey(
    'jwk',
    keyData,
    algorithm,
    extractable,
    usages
  );
}

function namedCurve(key: crypto.KeyObject): string {
  /* istanbul ignore next */
  switch (key.asymmetricKeyDetails?.namedCurve) {
    case 'prime256v1':
      return 'P-256';
    case 'secp384r1':
      return 'P-384';
    case 'secp521r1':
      return 'P-521';
    default:
      /* istanbul ignore next */
      throw new Error(
        `Unsupported curve ${key.asymmetricKeyDetails?.namedCurve}`
      );
  }
}
