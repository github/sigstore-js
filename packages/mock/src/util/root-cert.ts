/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { Crypto } from '@peculiar/webcrypto';
import x509 from '@peculiar/x509';

const MS_PER_DAY = 1000 * 60 * 60 * 24;

interface CertWithKey {
  cert: x509.X509Certificate;
  keyPair: CryptoKeyPair;
}

export async function createRootCertificate(
  name: string,
  keyPair: CryptoKeyPair,
  signAlgo: EcdsaParams
): Promise<CertWithKey> {
  const crypto = new Crypto();

  const tbs: x509.X509CertificateCreateSelfSignedParams = {
    serialNumber: '01',
    name: name,
    notBefore: new Date(),
    notAfter: new Date(Date.now() + 365 * MS_PER_DAY),
    signingAlgorithm: signAlgo,
    keys: keyPair,
    extensions: [
      new x509.BasicConstraintsExtension(true, undefined, true),
      new x509.KeyUsagesExtension(
        x509.KeyUsageFlags.cRLSign | x509.KeyUsageFlags.keyCertSign,
        true
      ),
      await x509.SubjectKeyIdentifierExtension.create(
        keyPair.publicKey,
        false,
        crypto
      ),
      await x509.AuthorityKeyIdentifierExtension.create(
        keyPair.publicKey,
        false,
        crypto
      ),
    ],
  };

  return {
    cert: await x509.X509CertificateGenerator.createSelfSigned(tbs, crypto),
    keyPair,
  };
}

export async function createIntermediateCertificate(
  subject: string,
  issuer: string,
  keyPair: CryptoKeyPair,
  signAlgo: EcdsaParams
): Promise<CertWithKey> {
  const crypto = new Crypto();

  const tbs: x509.X509CertificateCreateParams = {
    serialNumber: '02',
    subject,
    issuer,
    notBefore: new Date(),
    notAfter: new Date(Date.now() + 365 * MS_PER_DAY),
    signingAlgorithm: signAlgo,
    publicKey: keyPair.publicKey,
    signingKey: keyPair.privateKey,
    extensions: [
      new x509.BasicConstraintsExtension(false, undefined, true),
      new x509.KeyUsagesExtension(x509.KeyUsageFlags.digitalSignature, true),
      new x509.ExtendedKeyUsageExtension(
        [x509.ExtendedKeyUsage.timeStamping],
        true
      ),
      await x509.SubjectKeyIdentifierExtension.create(
        keyPair.publicKey,
        false,
        crypto
      ),
      await x509.AuthorityKeyIdentifierExtension.create(
        keyPair.publicKey,
        false,
        crypto
      ),
    ],
  };

  return {
    cert: await x509.X509CertificateGenerator.create(tbs, crypto),
    keyPair,
  };
}
