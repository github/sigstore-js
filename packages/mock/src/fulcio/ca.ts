/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { Crypto } from '@peculiar/webcrypto';
import x509 from '@peculiar/x509';
import * as asn1js from 'asn1js';
import type { KeyPairKeyObjectResult } from 'crypto';
import {
  DIGEST_SHA256,
  KEY_ALGORITHM_ECDSA_P256,
  SIGNING_ALGORITHM_ECDSA_SHA384,
} from '../constants';
import { keyObjectToCryptoKey } from '../util/key';
import { createRootCertificate } from '../util/root-cert';
import type { CTLog } from './ctlog';

const ISSUER = 'CN=sigstore,O=sigstore.mock';

const OID_SIGNED_CERTIFICATE_TIMESTAMP_LIST = '1.3.6.1.4.1.11129.2.4.2';

const MS_PER_MINUTE = 1000 * 60;

export interface CA {
  rootCertificate: Buffer;
  issueCertificate: (options: CertificateRequestOptions) => Promise<Buffer>;
}

export interface CertificateRequestOptions {
  publicKey: Buffer;
  subjectAltName: string;
  extensions?: ExtensionValue[];
}

interface ExtensionValue {
  oid: string;
  value: string;
  legacy?: boolean;
}

// Initialize the mock CA. Starts by generating a root key pair and self-signed
// certificate. The root certificate is used to sign any certificates issued by
// the #issueCertificate method. Doing this work in a function allows us to
// work around the fact that some of these operations are async and can't be
// done in the constructor.
export async function initializeCA(
  keyPair: KeyPairKeyObjectResult,
  ctLog?: CTLog,
  clock?: Date
): Promise<CA> {
  const cryptoKeyPair = {
    privateKey: await keyObjectToCryptoKey(keyPair.privateKey),
    publicKey: await keyObjectToCryptoKey(keyPair.publicKey),
  };

  const root = await createRootCertificate(
    ISSUER,
    cryptoKeyPair,
    SIGNING_ALGORITHM_ECDSA_SHA384
  );

  return new CAImpl({
    keyPair: root.keyPair,
    rootCertificate: root.cert,
    ctLog,
    clock,
  });
}

class CAImpl implements CA {
  private root: x509.X509Certificate;
  private keyPair: CryptoKeyPair;
  private ctLog?: CTLog;
  private getCurrentTime: () => Date;
  private crypto: Crypto;

  constructor(options: {
    rootCertificate: x509.X509Certificate;
    keyPair: CryptoKeyPair;
    ctLog?: CTLog;
    clock?: Date;
  }) {
    this.root = options.rootCertificate;
    this.ctLog = options.ctLog;
    this.keyPair = options.keyPair;
    this.getCurrentTime = () => options.clock || new Date();
    this.crypto = new Crypto();
  }

  public get rootCertificate(): Buffer {
    return Buffer.from(this.root.rawData);
  }

  public async issueCertificate(
    opts: CertificateRequestOptions
  ): Promise<Buffer> {
    // convert key to CryptoKey
    const key = await this.crypto.subtle.importKey(
      'spki',
      opts.publicKey,
      KEY_ALGORITHM_ECDSA_P256,
      true,
      ['sign', 'verify']
    );

    const now = this.getCurrentTime();

    // https://github.com/sigstore/fulcio/blob/549813ca04ae02b4a19181817a90a66c2a00960c/pkg/ca/common.go#L29
    const tbs = {
      serialNumber: '11182004',
      notBefore: now,
      notAfter: new Date(now.getTime() + 10 * MS_PER_MINUTE),
      issuer: ISSUER,
      publicKey: key,
      signingAlgorithm: SIGNING_ALGORITHM_ECDSA_SHA384,
      signingKey: this.keyPair.privateKey,
      extensions: [
        new x509.KeyUsagesExtension(x509.KeyUsageFlags.digitalSignature, true),
        new x509.ExtendedKeyUsageExtension([x509.ExtendedKeyUsage.codeSigning]),
        new x509.SubjectAlternativeNameExtension(
          [{ type: 'url', value: opts.subjectAltName }],
          true
        ),
        await x509.SubjectKeyIdentifierExtension.create(
          key,
          false,
          this.crypto
        ),
        await x509.AuthorityKeyIdentifierExtension.create(
          this.keyPair.publicKey,
          false,
          this.crypto
        ),
        ...generateExtensions(opts.extensions || []),
      ],
    } satisfies x509.X509CertificateCreateParams;

    if (this.ctLog) {
      // Pre-sign certificate for submission to CT Log
      const preCert = await x509.X509CertificateGenerator.create(
        tbs,
        this.crypto
      );

      // Submit to CT Log in exchange for SCT
      const issuerID = await this.issuerID;
      const sct = await this.ctLog.log(preCert.rawData, issuerID);

      // Add the SCT extension to the certificate
      tbs.extensions.push(extSCTList(sct));
    }

    const cert = await x509.X509CertificateGenerator.create(tbs, this.crypto);
    return Buffer.from(cert.rawData);
  }

  private get issuerID(): Promise<ArrayBuffer> {
    return this.crypto.subtle
      .exportKey('spki', this.keyPair.publicKey)
      .then((key) => this.crypto.subtle.digest({ name: DIGEST_SHA256 }, key));
  }
}

const extSCTList = (sct: ArrayBuffer): x509.Extension => {
  // Manually construct the SCTList
  const sctBuffer = sct;
  const sctList = Buffer.concat([
    Buffer.from([0x00]),
    Buffer.from([sctBuffer.byteLength]),
    Buffer.from(sctBuffer),
  ]);

  // Wrap the SCTList in an OctetString
  const sctListWrapper = new asn1js.OctetString({ valueHex: sctList });

  return new x509.Extension(
    OID_SIGNED_CERTIFICATE_TIMESTAMP_LIST,
    false,
    sctListWrapper.toBER(false)
  );
};

const generateExtensions = (extensions: ExtensionValue[]): x509.Extension[] =>
  extensions.map((extension) => {
    const value =
      extension.legacy === true
        ? Buffer.from(extension.value)
        : new asn1js.Utf8String({ value: extension.value }).toBER(false);
    return new x509.Extension(extension.oid, false, value);
  });
