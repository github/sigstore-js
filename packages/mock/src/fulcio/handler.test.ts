/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { generateKeyPairSync } from 'crypto';
import { generateKeyPair } from '../util/key';
import { CA, initializeCA } from './ca';
import { fulcioHandler } from './handler';

describe('fulcioHandler', () => {
  const keyPair = generateKeyPair('prime256v1');

  describe('#path', () => {
    it('returns the correct path', async () => {
      const ca = await initializeCA(keyPair);
      const handler = fulcioHandler(ca);
      expect(handler.path).toBe('/api/v2/signingCert');
    });
  });

  describe('#fn', () => {
    it('returns a function', async () => {
      const ca = await initializeCA(keyPair);
      const handler = fulcioHandler(ca);
      expect(handler.fn).toBeInstanceOf(Function);
    });

    describe('when invoked', () => {
      const { publicKey } = generateKeyPairSync('ec', { namedCurve: 'P-256' });

      const claims = {
        sub: 'http://github.com/foo/workflow.yml@refs/heads/main',
      };
      const jwt = jwtify(claims);

      const certRequest = {
        credentials: {
          oidcIdentityToken: jwt,
        },
        publicKeyRequest: {
          publicKey: {
            algorithm: 'ECDSA',
            content: publicKey
              .export({ format: 'pem', type: 'spki' })
              .toString('ascii'),
          },
          proofOfPossession: 'foobar',
        },
      };

      it('returns a certificate chain', async () => {
        const ca = await initializeCA(keyPair);
        const { fn } = fulcioHandler(ca);

        // Make a request
        const resp = await fn(JSON.stringify(certRequest));
        expect(resp.statusCode).toBe(201);

        // Check the response
        const certs = JSON.parse(resp.response.toString());
        expect(certs).toBeDefined();
        expect(certs.signedCertificateEmbeddedSct).toBeDefined();
        expect(certs.signedCertificateEmbeddedSct.chain).toBeDefined();
        expect(
          certs.signedCertificateEmbeddedSct.chain.certificates
        ).toBeDefined();
        expect(
          certs.signedCertificateEmbeddedSct.chain.certificates
        ).toHaveLength(2);
      });

      describe('when the CA raises an error', () => {
        const ca: CA = {
          rootCertificate: Buffer.from(''),
          issueCertificate: async () => {
            throw new Error('oops');
          },
        };

        it('returns 400 error', async () => {
          const { fn } = fulcioHandler(ca, { strict: false });
          const resp = await fn(JSON.stringify(certRequest));
          expect(resp.statusCode).toBe(400);
        });
      });
    });
  });
});

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
function jwtify(obj: any): string {
  return `.${Buffer.from(JSON.stringify(obj)).toString('base64')}.`;
}
