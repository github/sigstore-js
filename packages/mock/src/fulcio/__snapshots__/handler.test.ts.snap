// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fulcioHandler #fn when invoked returns a certificate chain 1`] = `
Extensions [
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.1
  OCTET STRING : 687474703a2f2f666f6f2e636f6d",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.8
  OCTET STRING :
    UTF8String : 'http://foo.com'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.2
  OCTET STRING : 776f726b666c6f775f6469737061746368",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.20
  OCTET STRING :
    UTF8String : 'workflow_dispatch'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.9
  OCTET STRING :
    UTF8String : 'https://github.com/foo/attest-demo/.github/workflows/oidc.yml@refs/heads/main'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.10
  OCTET STRING :
    UTF8String : 'ba214227977e57973d87219493ad93eeda9c7d6c'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.6
  OCTET STRING : 726566732f68656164732f6d61696e",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.14
  OCTET STRING :
    UTF8String : 'refs/heads/main'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.5
  OCTET STRING : 666f6f2f6174746573742d64656d6f",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.12
  OCTET STRING :
    UTF8String : 'https://github.com/foo/attest-demo'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.15
  OCTET STRING :
    UTF8String : '792829709'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.16
  OCTET STRING :
    UTF8String : 'https://github.com/foo'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.17
  OCTET STRING :
    UTF8String : '398027'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.22
  OCTET STRING :
    UTF8String : 'public'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.11
  OCTET STRING :
    UTF8String : 'github-hosted'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.3
  OCTET STRING : 62613231343232373937376535373937336438373231393439336164393365656461396337643663",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.13
  OCTET STRING :
    UTF8String : 'ba214227977e57973d87219493ad93eeda9c7d6c'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.4
  OCTET STRING : 4f494443",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.18
  OCTET STRING :
    UTF8String : 'https://github.com/foo/attest-demo/.github/workflows/oidc.yml@refs/heads/main'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.19
  OCTET STRING :
    UTF8String : 'ba214227977e57973d87219493ad93eeda9c7d6c'",
  "SEQUENCE :
  OBJECT IDENTIFIER : 1.3.6.1.4.1.57264.1.21
  OCTET STRING :
    UTF8String : 'https://github.com/foo/attest-demo/actions/runs/11997537386/attempts/3'",
]
`;
